File: CodingTest/성호/21주차/등굣길.js
function solution(m, n, puddles) {
  let dp = Array.from({ length: n }, () => Array(m).fill(0));

  for (let [x, y] of puddles) {
    dp[y - 1][x - 1] = -1;
  }
  dp[0][0] = 1;
  for (let i = 0; i < n; i++) {
    for (let j = 0; j < m; j++) {
      if (dp[i][j] === -1) {
        dp[i][j] = 0;
        continue;
      }
      if (i !== 0) {
        dp[i][j] += dp[i - 1][j] % 1000000007;
      }
      if (j !== 0) {
        dp[i][j] += dp[i][j - 1] % 1000000007;
      }
    }
  }
  return dp[n - 1][m - 1] % 1000000007;
}

File: CodingTest/한종민/22주차/등굣길.java
import java.util.*;

class Solution {
    public int solution(int m, int n, int[][] puddles) {
        int[][] dp = new int[n][m];
        boolean[][] isPuddle = new boolean[n][m];

        // 물웅덩이 위치 표시
        for (int[] puddle : puddles) {
            int y = puddle[1] - 1;
            int x = puddle[0] - 1;
            isPuddle[y][x] = true;
        }

        dp[0][0] = 1; // 시작점

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (isPuddle[i][j]) {
                    dp[i][j] = 0;
                    continue;
                }

                if (i > 0) dp[i][j] += dp[i - 1][j];
                if (j > 0) dp[i][j] += dp[i][j - 1];

                dp[i][j] %= 1_000_000_007;
            }
        }

        return dp[n - 1][m - 1];
    }
}
